### PR TITLE
[GERAPUMP-22] Merge akka-streams branch into master

### PR DESCRIPTION
**Pull request by manuzhang**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the commit message is formatted like:
   `[GEARPUMP-<Jira issue #>] Meaningful description of pull request` 
 - [x] Make sure tests pass via `sbt clean test`.
 - [x] Make sure old documentation affected by the pull request has been updated and new documentation added for new functionality.

### CODE REVIEW

File: project/BuildExperiments.scala [hunk 1]
```diff
@@ -25,7 +25,7 @@ import sbt.Keys._
 object BuildExperiments extends sbt.Build {
 
   lazy val experiments: Seq[ProjectReference] = Seq(
-    // akkastream,
+    akkastream,
```
Comment (manuzhang) [inline project/BuildExperiments.scala]: now build is activated for `akkastream` project

File: project/Dependencies.scala [hunk 2]
```diff
@@ -23,7 +23,8 @@ object Dependencies {
 
   val crossScalaVersionNumbers = Seq("2.11.8")
   val scalaVersionNumber = crossScalaVersionNumbers.last
-  val akkaVersion = "2.4.3"
+  val akkaVersion = "2.4.16"
```
Comment (manuzhang) [inline project/Dependencies.scala]: Note that Akka version is updated to **2.4.16**

File: project/Dependencies.scala [hunk 3]
```diff
@@ -23,7 +23,8 @@ object Dependencies {
...
+  val akkaHttpVersion = "10.0.1"
```
Comment (manuzhang) [inline project/Dependencies.scala]: Note that akka-http version is updated to **10.0.1**

File: streaming/src/main/scala/org/apache/gearpump/streaming/StreamApplication.scala [hunk 4]
```diff
@@ -123,7 +123,7 @@ object LifeTime {
  */
 class StreamApplication(
     override val name: String, val inputUserConfig: UserConfig,
-    dag: Graph[ProcessorDescription, PartitionerDescription])
+    val dag: Graph[ProcessorDescription, PartitionerDescription])
```
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/StreamApplication.scala]: We need to access `dag` in `RemoteMaterializerImpl` for akka-streams.

**Comment by manuzhang** [comment]
R: @huafengw Most changes are confined to akka-streams module while the major exception is upgrading akka version. I've made some inline comments to note it.

**Comment by huafengw** [comment]
+1

**Pull request closed without merging**